service/s3/s3crypto/s3crypto_client.go

```diff
@@ -0,0 +1,175 @@
+	MasterKey []byte
+}
+
+// New instantiates a new S3 crypto client
+// Example:
+// cmkID := "some key id to kms"
+// kp, err = s3crypto.NewKMSKeyProvider(session.New(), cmkID, s3crypto.NewJSONMatDesc())
+// if err != nil {
+//   return err
+// }
+// svc := s3crypto.New(s3crypto.Authentication(kp))
+func New(mode CryptoMode, options ...func(*Client)) *Client {
+	sess := session.New()
+	client := &Client{
+		Config: Config{
+			Mode:         mode,
+			SaveStrategy: NewHeaderSaveStrategy(),
+			KMSSession:   sess,
+			S3Session:    sess,
+		},
+	}
+
+	for _, option := range options {
+		option(client)
+	}
+
+	client.S3 = s3.New(client.Config.S3Session)
+	return client
+}
+
+// PutObjectRequest creates a temp file to encrypt the contents into. It then streams
+// that data to S3.
+//
+// Example:
+// svc := s3crypto.New(s3crypto.Authentication(kp))
+// req, out := svc.PutObjectRequest(&s3.PutObjectInput {
+//   Key: aws.String("testKey"),
+//   Bucket: aws.String("testBucket"),
+//   Body: bytes.NewBuffer("test data"),
+// })
+// err := req.Send()
+func (c *Client) PutObjectRequest(input *s3.PutObjectInput) (*request.Request, *s3.PutObjectOutput) {
+	req, out := c.S3.PutObjectRequest(input)
+
+	// Create temp file to be used later for calculating the SHA256 header
+	f, err := ioutil.TempFile("./", "")
+	if err != nil {
+		req.Error = err
+		return req, out
+	}
+
+	req.Handlers.Build.PushFront(func(r *request.Request) {
+		md5 := newMD5Reader(input.Body)
+		sha := newSHA256Writer(f)
+		err = c.Config.Mode.EncryptContents(sha, md5)
+		if err != nil {
+			r.Error = err
+			return
+		}
+
+		shaHex := hex.EncodeToString(sha.GetValue())
+		req.HTTPRequest.Header.Set("X-Amz-Content-Sha256", shaHex)
+
+		f.Seek(0, 0)
+		input.Body = f
+		log.Println(input)
+
+		env, err := EncodeMeta(md5, c.Config.Mode)
+		if err != nil {
+			r.Error = err
+			return
+		}
+		err = c.Config.SaveStrategy.Save(env, input)
+		r.Error = err
+	})
+
+	fn := func(r *request.Request) {
+		// Close the temp file and cleanup
+		f.Close()
+		os.Remove(f.Name())
+	}
+	req.Handlers.Send.PushBack(fn)
+	req.Handlers.ValidateResponse.PushBack(fn)
+	return req, out
+}
+
+// PutObject is a wrapper for PutObjectRequest
+func (c *Client) PutObject(input *s3.PutObjectInput) (*s3.PutObjectOutput, error) {
+	req, out := c.PutObjectRequest(input)
+	return out, req.Send()
+}
+
+// GetObjectRequest will make a request to s3 and retrieve the object. In this process
+// decryption will be done. We only support V2 reads of KMS and GCM.
+//
+// Example:
+// svc := s3crypto.New(s3crypto.Authentication(kp))
+// req, out := svc.GetObjectRequest(&s3.GetObjectInput {
+//   Key: aws.String("testKey"),
+//   Bucket: aws.String("testBucket"),
+// })
+// err := req.Send()
+func (c *Client) GetObjectRequest(input *s3.GetObjectInput) (*request.Request, *s3.GetObjectOutput) {
+	req, out := c.S3.GetObjectRequest(input)
+	req.Handlers.Unmarshal.PushBack(func(r *request.Request) {
```

Comment:
If this handler fails should the `out.Body` be closed? since users may not close a body if an API call fails. Generally they'd only close the body if the API call was successful.